bundle/src/main/java/com/adobe/acs/commons/indesign/dynamicdeckdynamo/services/impl/XMLGeneratorServiceImpl.java

```diff
@@ -0,0 +1,301 @@
+
+import com.adobe.acs.commons.indesign.dynamicdeckdynamo.constants.DynamicDeckDynamoConstants;
+import com.adobe.acs.commons.indesign.dynamicdeckdynamo.exception.DynamicDeckDynamoException;
+import com.adobe.acs.commons.indesign.dynamicdeckdynamo.pojos.XMLResourceIterator;
+import com.adobe.acs.commons.indesign.dynamicdeckdynamo.services.XMLGeneratorService;
+import com.adobe.acs.commons.indesign.dynamicdeckdynamo.utils.DynamicDeckUtils;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.dam.api.AssetManager;
+import com.day.cq.dam.api.DamConstants;
+import com.day.cq.dam.commons.util.DamUtil;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.jackrabbit.webdav.bind.BindConstants;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ValueMap;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Result;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.*;
+
+/**
+ * This class is used to parse Annotated Template XML and generate Indd XML
+ */
+@Component(service = XMLGeneratorService.class)
+public class XMLGeneratorServiceImpl implements XMLGeneratorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(XMLGeneratorServiceImpl.class);
+
+    /**
+     * This Method Parses an Annotated Template XML and Creates an INDD XML out of the Template. The INDD XML is sent to
+     * INDD Server for generation of decks.
+     */
+    @Override
+    public String generateInddXML(InputStream xmlInputStream,
+                                  List<XMLResourceIterator> assetItrList, Resource masterResource, Resource deckResource,
+                                  ResourceResolver resourceResolver, List<String> imageList) throws DynamicDeckDynamoException {
+
+        InputStream resultInputStream;
+        Asset processXmlAsset = null;
+        final String xmlName = "processedXml.xml";
+
+        if (null == masterResource) {
+            masterResource = deckResource;
+        }
+
+        File targetFile = new File("targetFile-" + Calendar.getInstance().getTimeInMillis() + DynamicDeckDynamoConstants.DASH
+                + StringUtils.substringBeforeLast(xmlName, DynamicDeckDynamoConstants.DOT) + ".tmp");
```

Comment:
You can create temporary files using ```File.createTempoFile``` (java.util.File).

Reply:
It also seems unnecessary to have a file here. I could be wrong, but AFAIK, the concept here is that there is an `InputStream` (`xmlInputStream`) which then gets parsed into a DOM object, processed (`populateDocNodes`) and then serialized back to XML. You should just be able to parse the input stream directly without storing it in a file.

OTOH, where it might be useful to use a temporary file is to store the serialized output. That would avoid the memory issue @joerghoh mentioned.

Reply:
what's a realistic range for the side of these XML files? I've not done much work in this space so have not sure if we're talking a KBs, MBs, or more!

Reply:
As I explained, you don't need to use a temp file here.

Reply:
Also, it isn't really clear to me that this temporary file is cleaned up. Again, it can be removed in this context, but any temporary file should:
* Be created using `File.createTempFile` (this way you ensure that the file is created in the correct directory -- this case creates a file in the current working directory which wouldn't be expected)
* Be cleaned up -- `File.deleteOnExit()` is OK for this on Cloud Service (since the JVMs are restarted daily), but for Quickstart-based deployments it would be better to manually do that cleanup since AEM JVM processes tend to be long-lived and you could build up quite a few of these files.